kratos/utilities/function_parser_utility.cpp

```diff
@@ -0,0 +1,184 @@
+
+GenericFunctionUtility::~GenericFunctionUtility()
+{
+    te_free(mpTinyExpr);
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+std::string GenericFunctionUtility::FunctionBody()
+{
+    std::string aux_string = mFunctionBody;
+    std::unordered_map<std::string,std::string> aux_replace_letters({{"**","^"},{"X","x0"},{"Y","y0"},{"Z","z0"}});
+    for (auto& r_pair : aux_replace_letters) {
+         aux_string = StringUtilities::ReplaceAllSubstrings(aux_string, r_pair.second, r_pair.first);
+    }
+    return aux_string;
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+double GenericFunctionUtility::RotateAndCallFunction(
+    const double x,
+    const double y,
+    const double z,
+    const double t,
+    const double X,
+    const double Y,
+    const double Z
+    )
+{
+    array_1d<double,3> xglobal;
+    xglobal[0] = x;
+    xglobal[1] = y;
+    xglobal[2] = z;
+    array_1d<double,3> xlocal = prod(mRotationMatrix, (xglobal - mCenterCoordinates) );
+    array_1d<double,3> xglobal_initial;
+    xglobal_initial[0] = X;
+    xglobal_initial[1] = Y;
+    xglobal_initial[2] = Z;
+    array_1d<double,3> xlocal_initial = prod(mRotationMatrix, (xglobal_initial - mCenterCoordinates) );
+    return CallFunction(xlocal[0],xlocal[1],xlocal[2], t,xlocal_initial[0],xlocal_initial[1],xlocal_initial[2]);
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+double GenericFunctionUtility::CallFunction(
+    const double x,
+    const double y,
+    const double z,
+    const double t,
+    const double X,
+    const double Y,
+    const double Z
+    )
+{
+    mNameSpace["x"] = x;
+    mNameSpace["y"] = y;
+    mNameSpace["z"] = z;
+    mNameSpace["X"] = X;
+    mNameSpace["Y"] = Y;
+    mNameSpace["Z"] = Z;
+    mNameSpace["t"] = t;
+
+    return te_eval(mpTinyExpr);
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void GenericFunctionUtility::InitializeParser()
+{
+    // Initialize
+    if (mpTinyExpr == nullptr) {
+        int err;
+
+        /* Defining table */
+        double& x = mNameSpace["x"];
+        double& y = mNameSpace["y"];
+        double& z = mNameSpace["z"];
+        double& t = mNameSpace["t"];
+        double& X = mNameSpace["X"];
+        double& Y = mNameSpace["Y"];
+        double& Z = mNameSpace["Z"];
+
+        /* Store variable names and pointers. */
+        const te_variable vars[] = {{"x", &x}, {"y", &y}, {"z", &z}, {"t", &t}, {"x0", &X}, {"y0", &Y}, {"z0", &Z}};
```

Comment:
here you get the variables from the mNameSpace, this is an overhead which we could avoid

Reply:
Please look at CallFunction()